File: plot_stability.py
import numpy as np
import sys
import matplotlib
import matplotlib.pyplot as plt

if __name__ == '__main__':
    if len(sys.argv) != 2:
        print("Usage:\n\tpython {} <method name>\n".format(sys.argv[0]))
        print("where <method name> is either forward_euler or crank_nicolson")
        exit(1)

    method = sys.argv[1]

    coefficients = ['a_1', 'a_2', 'a_3']

    N = 127
    dxSquared = 1.0 / (N+1)**2

    timesteps = 0.5*dxSquared * np.array([128., 8., 1.])
    fig, axes = plt.subplots(len(coefficients), len(timesteps))

    for i, coefficient in enumerate(coefficients):
        for j, timestep in enumerate(timesteps):
            filename = '{method}_stability_u_a{i}_dt{j}.txt'.format(
                i=i, j=j, method=method)

            u = np.loadtxt(filename)
            x = np.linspace(0, 1, u.shape[1])

            axes[i,j].plot(x, u[-1,:])
            axes[i,j].set_title('$a_{i}$, $\\Delta t_{j}$'.format(i=i+1,
                                                                  j=j+1))
    fig.tight_layout()
    plt.savefig('stability_{}.png'.format(method))
    plt.show()

File: plot_stability.m
function plot_stability(method)

    coefficients = {'a_1', 'a_2', 'a_3'};

    N = 127;
    dxSquared = 1.0 / (N+1)^2;

    timesteps = 0.5*dxSquared * [128., 8., 1.];
    

    for i = 1:3
        
        for j = 1:3
            filename = sprintf('%s_stability_u_a%d_dt%d.txt',method, i-1, j-1);

            u = load(filename);
            x = linspace(0, 1, size(u, 2));

            subplot(3, 3, 3*(i-1) + j);
            plot(x, u(end,:));
            title(sprintf('a_{%d}, \\Delta t_{%d}', i, j));
        end;
    end;
    
       
